content/hoon/lagoon/conv.md

```diff
@@ -0,0 +1,962 @@
+  ?>  =(+((lent dex)) (lent shape))
+  =/  res
+    %-  zeros
+    :*  ~[1 (snag 0 (flop shape))]
+        bloq
+        kind
+        ~
+    ==
+  =/  idx  0
+  |-  ^-  ray
+  ?:  =((snag 0 (flop shape.meta.res)) idx)  res
+  =/  val  (get-item a (weld dex ~[idx]))
+  $(idx +(idx), res (set-item res ~[0 idx] val))
+```
+
+---
+
+## `+set-row` {#setrow}
+
+Set row at index in a `$ray`.
+
+#### Accepts
+
+A `$ray`, target index as a list of multi-dimensional indices, and the row to set.
+
+#### Produces
+
+The updated `$ray`.
+
+#### Source
+
+```hoon
+++  set-row
+  |=  [a=ray dex=(list @) row=ray]
+  ^-  ray
+  ?:  &(=(1 (lent dex)) =(1 (lent shape.meta.row)))  (set-item a dex (get-item row ~[0]))
+  ?>  =(+((lent dex)) (lent shape.meta.a))
+  ?>  =((lent shape.meta.row) 2)
+  ?>  =(1 (snag 0 shape.meta.row))
+  ?>  =((snag 1 shape.meta.row) (snag 0 (flop shape.meta.a)))
+  =/  idx  0
+  |-  ^-  ray
+  ?:  =((snag 1 shape.meta.row) idx)  a
+  %=  $
+    idx  +(idx)
+    a    (set-item a (weld dex ~[idx]) (get-item row ~[0 idx]))
+  ==
+```
+
+---
+
+## `+get-col` {#getcol}
+
+Get column at index from a `$ray`.
+
+#### Accepts
+
+A `$ray` and target index as a single-entry list of multi-dimensional indices.
+
+#### Produces
+
+A `$ray`. of dimension $n-1$.
+
+#### Source
+
+```hoon
+++  get-col
+  |=  [a=ray dex=(list @)]
+  ^-  ray
+  (get-row (transpose a) dex)
+```
+
+---
+
+## `+set-col` {#setcol}
+
+Set column at index in a `$ray`.
+
+#### Accepts
+
+A `$ray`, target index as a list of multi-dimensional indices, and the column to set.
+
+#### Produces
+
+The updated `$ray`.
+
+#### Source
+
+```hoon
+++  set-col
+  |=  [a=ray dex=(list @) col=ray]
+  ^-  ray
+  (transpose (set-row (transpose a) dex col))
+```
+
+---
+
+## `+get-bloq-offset` {#getbloqoffset}
+
+Utility function to get the bloq offset of an n-dimensional index.
+
+#### Accepts
+
+A `$ray` and target index as a list of multi-dimensional indices.
+
+#### Produces
+
+A `@`
+
+#### Source
+
+```hoon
+++  get-bloq-offset  ::  get bloq offset of n-dimensional index
+  |=  [=meta dex=(list @)]
+  ^-  @
+  (get-item-number shape.meta dex)
+```
+
+---
+
+## `+get-item-number` {#getitemnumber}
+
+Utility function to convert an n-dimensional index to a scalar index.
+
+#### Accepts
+
+Parameters as specified in source
+
+#### Produces
+
+An index `@`.
+
+#### Source
+
+```hoon
+++  get-item-number  ::  convert n-dimensional index to scalar index
+  |=  [sap=(list @) dex=(list @)]
+  ^-  @
+  =/  cof  1
+  =/  ret  0
+  =.  sap  (flop sap)
+  =.  dex  (flop dex)
+  |-  ^+  ret
+  ?~  sap  ret :: out of indices, return
+  ?~  dex  !!  :: no indices past size
+  ?>  (^lth i.dex i.sap)  :: index must be less than size
+  %=  $
+    sap  t.sap
+    dex  t.dex
+    cof  (^mul cof i.sap)
+    ret  (^add ret (^mul i.dex cof))
+  ==
+```
+
+---
+
+## `+strides` {#strides}
+
+Return the stride in each dimension:  row, col, layer, &c. The stride is reported in units of bits.
```

Comment:
First sentence is a bit too informal for me, should be "Return the stride in each dimension: row, column, layer, etc."